src/server/task_manager/task_store.ts

```diff
@@ -0,0 +1,363 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * This module contains helpers for managing the task manager storage layer.
+ */
+
+import { ConcreteTaskInstance, ElasticJs, TaskInstance, TaskStatus } from './task';
+
+const DOC_TYPE = '_doc';
+
+export interface StoreOpts {
+  callCluster: ElasticJs;
+  index: string;
+  maxAttempts: number;
+  supportedTypes: string[];
+}
+
+export interface FetchOpts {
+  searchAfter?: any[];
+  sort?: object[];
+  query?: object;
+}
+
+export interface FetchResult {
+  searchAfter: any[];
+  docs: ConcreteTaskInstance[];
+}
+
+export interface RemoveResult {
+  index: string;
+  id: string;
+  version: string;
+  result: string;
+}
+
+// Internal, the raw document, as stored in the Kibana index.
+export interface RawTaskDoc {
+  _id: string;
+  _index: string;
+  _type: string;
+  _version: number;
+  _source: {
+    type: string;
+    task: {
+      taskType: string;
+      runAt: Date;
+      interval?: string;
+      attempts: number;
+      status: TaskStatus;
+      params: string;
+      state: string;
+      user?: string;
+      scope?: string | string[];
+    };
+  };
+}
+
+/**
+ * Wraps an elasticsearch connection and provides a task manager-specific
+ * interface into the index.
+ */
+export class TaskStore {
+  private callCluster: ElasticJs;
+  private index: string;
+  private maxAttempts: number;
+  private supportedTypes: string[];
+
+  /**
+   * Constructs a new TaskStore.
+   * @param {StoreOpts} opts
+   * @prop {CallCluster} callCluster - The elastic search connection
+   * @prop {string} index - The name of the task manager index
+   * @prop {number} maxAttempts - The maximum number of attempts before a task will be abandoned
+   * @prop {string[]} supportedTypes - The task types supported by this store
+   */
+  constructor(opts: StoreOpts) {
+    this.callCluster = opts.callCluster;
+    this.index = opts.index;
+    this.maxAttempts = opts.maxAttempts;
+    this.supportedTypes = opts.supportedTypes;
+  }
+
+  /**
+   * Initializes the store, ensuring the task manager index is created and up to date.
+   */
+  public async init() {
+    const properties = {
+      type: { type: 'keyword' },
+      task: {
+        properties: {
+          taskType: { type: 'keyword' },
+          runAt: { type: 'date' },
+          interval: { type: 'text' },
+          attempts: { type: 'integer' },
+          status: { type: 'keyword' },
+          params: { type: 'text' },
+          state: { type: 'text' },
+          user: { type: 'keyword' },
+          scope: { type: 'keyword' },
+        },
+      },
+    };
+
+    try {
+      await this.callCluster('indices.create', {
```

Comment:
Let's make sure the settings of this index match other system indices. Should be something like:
```
    "settings": {
      "index": {
        "auto_expand_replicas": "0-1",
        "codec": "best_compression",
        "format": "6",
        "number_of_replicas": "0",
        "number_of_shards": "1"
      }
    },
```

Using auto expand replicas makes sure the cluster will not be in yellow state because of unassigned replicas of this index.

Reply:
Chatted with @chrisdavies , specifying `number_of_shards` and `auto_expand_replicas` should be enough